Automattic/wp-calypso
# Session Restore: Move to pre-section routing

If successful, supersedes #16291. Main advantages of this approach:
- If there is a session to restore, we skip loading the Reader module altogether. The visual glitch is also gone, and performance is obviously better.
- The Reader module is left untouched and is unaware of this mechanism.
- No virtual Calypso section needs to be cooked up for this.

This PR _cheats_ by using localStorage instead of localforage (Session Restore's MVP used localStorage until #14196).

See commit messages for more insight.

**Question:** localforage is asynchronous, so to use it we'd have to refactor `server/bundler/loader` accordingly, which is something I would consider avoiding. Maybe `localStorage` could be good enough for this?

**Contributing:** If you are unfamiliar with `server/bundler/loader`, know that every change there will require a full Calypso rebuild to effect.

**Figure 1:** With no saved path to restore

![1](https://user-images.githubusercontent.com/150562/28635086-44f05f20-7231-11e7-888e-2f10c88ad304.gif)

**Figure 2:** With a saved path to restore

![2](https://user-images.githubusercontent.com/150562/28635093-4850f09e-7231-11e7-918c-0b9f794592bf.gif)

## TO TEST

**NOTE:** Feel free to substitute http://calypso.localhost:3000/ for the following if you prefer to run this branch locally

* Start here: https://calypso.live/?branch=fix/path-restore-earlier
* Click around -- go wild!
  * Make sure you can browse to the Reader as before
* Close the tab (and / or open an additional one)
* Start back at https://calypso.live
* Note where you land -- is it expected & delightful?
* Repeat the above, but end your "session" at `/` (the Reader root)
  * Do you land at the `/` Reader when you return?
* Clear your localStorage (`localStorage.clear()` in the console or log out and back in)
* Browse to https://calypso.live
  * Do you land at the `/` Reader as expected?
* Try to break stuff!

### Test with the config flag disabled:

* Run locally with the `restore-last-location` flag disabled & the `no-force-sympathy` forced on to make sure your local storage doesn't get cleared out:
  `ENABLE_FEATURES=no-force-sympathy DISABLE_FEATURES=restore-last-location npm start`
* Browse to http://calypso.localhost:3000/
* Whitelist the debugger scope in your console:
  * `localStorage.setItem( 'debug', 'calypso:restore-last-location' )`
* Click around
  * You should not see debug messages related to this feature
* Browse somewhere other than `/` & close the tab
* Browse to http://calypso.localhost:3000/
  * You should land at the `/` Reader (i.e. this feature is decidedly not active)

## TODO before merging
- [x] Update the [middleware README](https://github.com/Automattic/wp-calypso/blob/578c6b776330658ed6948cc8eb0732bfe26d8568/client/state/routing/README.md) to indicate the change in code structure here
- [x] Add a README to the [restore-last-path lib](https://github.com/Automattic/wp-calypso/tree/578c6b776330658ed6948cc8eb0732bfe26d8568/client/lib/restore-last-path)
